File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*,Node*>dict1;
        Node* curr = head;
        while(curr){
            Node* temp = new Node(curr->val);
            dict1[curr] = temp;
            curr = curr->next;
        }
        curr = head;
        while(curr){
            dict1[curr]->next = dict1[curr->next];
            dict1[curr]->random = dict1[curr->random];
            curr = curr->next;
        }
        return dict1[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        dict1 = {None:None}
        curr = head
        while curr:
            temp = Node(curr.val)
            dict1[curr] = temp
            curr = curr.next
        curr = head
        while curr:
            temp = dict1[curr]
            temp.next = dict1[curr.next]
            temp.random = dict1[curr.random]
            curr = curr.next
        return dict1[head]
            
            
        